### DISCUSSION TITLE
Is that possible to pass parameters to view collection?

### DISCUSSION THREAD

**Discussion by iamdineshkumar** [pocketbase/pocketbase#1948] [Q&A]
While making query to view collection whether able to pass values which can be utilized in query. For an instance we need to sum and return sales amount for particular period, we can pass that period so can make that query dynamic.

**Comment by ganigeorgiev** [comment]
It depends what do you understand by "pass parameters".

Similar as the other collection types, you can add additional conditions to the list result using the `fiter` query parameter but aggregations are not supported and you can filter only the already resolved view records.

Views are stored SQL statements and they usually don't support dynamic parameters.

As a workaround, if a simple `filter` expression is not enough and if the aggregation periods are fixed, you can always create individual views for each period.

**Reply by iamdineshkumar** [reply]
To make a confirmation, If I write following In Query,
Select (ROW_NUMBER() OVER()) as id, sum(saleamount) as amount from sales

And I can use filter query with date from and to add and it will add by pocketbase to above statement.
Am I right?

**Reply by ganigeorgiev** [reply]
I'm not sure if this is what you mean by "it will add by pocketbase to above statement", but the `filter` conditions are applied **on top of the resolved view collection records**. The view query itself can't be parameterized.

**Reply by iamdineshkumar** [reply]
I mean following

If I am using following in client end
const resultList = await pb.collection('saleValue').getList(1, 50, {
    filter: 'saleDate >= "2022-01-01 00:00:00" && saleDate < "2022-01-10 00:00:00"',
});

Then pocketbase will return result for 
Select (ROW_NUMBER() OVER()) as id, sum(saleamount) as amount from sales where saleDate >= "2022-01-01 00:00:00" && saleDate < "2022-01-10 00:00:00"

**Reply by ganigeorgiev** [reply]
No, this is not how it works. The client-side `filter` operates **on top of the resolved view collection records**, it doesn't append the extra conditions to the view query. The view query cannot be dynamic.

**Reply by iamdineshkumar** [reply]
Got it... But this option is available which will be helpful in certain cases as I mentioned.

**Reply by ganigeorgiev** [reply]
As mentioned views are stored SQL statements and they normally don't support dynamic parameters (I'm not aware of db that support parameterized views).

What you want is to execute dynamic raw sql aggregate statements from client-side and this is not supported because it is very difficult to generalize (_eg. what if there is a `GROUP BY`? should any extra conditions be applied to the `WHERE` or `HAVING` clause?, etc._) and do it securely.

If you want a parameterized dynamic custom query, you'll have to [register a custom route](https://pocketbase.io/docs/router/#register-new-route).

**Reply by iamdineshkumar** [reply]
Is that possible to have date field in view collection for an eg
''' 
SELECT
  (ROW_NUMBER() OVER()) as id,
STRFTIME("%Y-%m-%d", sale_on) as sale_date,
  count(sale_amount) as total_sale
FROM daily_sale
GROUP BY STRFTIME("%d-%m-%Y", sale_on)
'''

**Reply by ganigeorgiev** [reply]
@iamdineshkumar I'm not sure that I understand what do you mean.
`strftime` should work fine if `sale_on` is  a field in `daily_sale`.
It will be detected as a `json` field since `strftime` could yield to a datetime layout that is not supported by the `date` field. But in the end it shouldn't really matter since it will be just a string anyway.

**Reply by iamdineshkumar** [reply]
I tried to cast as datetime but cast is not allowing, getting error msg as 0:Must be in a valid format
'''
SELECT
(ROW_NUMBER() OVER()) as id,
CAST(STRFTIME("%Y-%m-%d", sale_on) as INT) as sale_date,
count(sale_amount) as total_sale
FROM daily_sale
GROUP BY STRFTIME("%d-%m-%Y", sale_on)
'''

As well as I have tried
'''
SELECT
(ROW_NUMBER() OVER()) as id,
CAST(strftime('%s', strftime('%Y-%m-%dT00:00:00+00:00', sale_on), 'unixepoch') as datetime) as sale_date,
count(sale_amount) as total_sale
FROM daily_sale
GROUP BY STRFTIME("%d-%m-%Y", sale_on)
'''

**Reply by iamdineshkumar** [reply]
atleast if I am getting sale_Date in epoch time stamp or in date time in view collection then it will be easy to filter details by query

**Reply by ganigeorgiev** [reply]
I'm sorry, I still don't understand what do you mean and what you are trying to do.
Why do you need the extra `CAST` in the first place (_it also works fine for me and I'm not getting the "Must be in a valid format" error_)?
The result value of strftime is a string and you can filter it the same way as a `date` field, aka. `sale_date >= "2023-03-14"`
